banyan/src/forest/mod.rs

```diff
@@ -0,0 +1,281 @@
+//! creation and traversal of banyan trees
+use super::index::*;
+use crate::store::{BlockWriter, ReadOnlyStore};
+use anyhow::Result;
+use core::{fmt::Debug, hash::Hash, iter::FromIterator, marker::PhantomData, ops::Range};
+use futures::future::BoxFuture;
+use libipld::{cbor::DagCbor, Ipld};
+use lru::LruCache;
+use rand::RngCore;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    fmt::Display,
+    sync::{Arc, Mutex},
+};
+mod read;
+mod stream;
+mod write;
+
+pub type FutureResult<'a, T> = BoxFuture<'a, Result<T>>;
+
+#[derive(Debug, Clone)]
+pub struct BranchCache<T: TreeTypes>(Arc<Mutex<lru::LruCache<T::Link, Branch<T>>>>);
+
+impl<T: TreeTypes> BranchCache<T> {
+    pub fn new(capacity: usize) -> Self {
+        Self(Arc::new(Mutex::new(LruCache::new(capacity))))
+    }
+
+    pub fn get<'a>(&'a self, link: &'a T::Link) -> Option<Branch<T>> {
+        self.0.lock().unwrap().get(link).cloned()
+    }
+
+    pub fn put(&self, link: T::Link, branch: Branch<T>) {
+        self.0.lock().unwrap().put(link, branch);
+    }
+}
+
+/// Trees can be parametrized with the key type and the sequence type. Also, to avoid a dependency
+/// on a link type with all its baggage, we parameterize the link type.
+///
+/// There might be more types in the future, so this essentially acts as a module for the entire
+/// code base.
+pub trait TreeTypes: Debug + Send + Sync + Clone + 'static {
+    /// key type. This also doubles as the type for a combination (union) of keys
+    type Key: Debug + Eq + Send;
+    /// compact sequence type to be used for indices
+    type Seq: CompactSeq<Item = Self::Key>
+        + Serialize
+        + DeserializeOwned
+        + Clone
+        + Debug
+        + FromIterator<Self::Key>
+        + Send
+        + Sync;
+    /// link type to use over block boundaries
+    type Link: Display + Debug + Hash + Eq + Clone + Copy + Send + Sync + DagCbor;
+
+    fn serialize_branch(links: Vec<(usize, Ipld)>, data: Vec<u8>) -> anyhow::Result<Vec<u8>>;
```

Comment:
Why not `fn serialize_branch(links: BTreeMap<usize, Cid>, data: Vec<u8>) -> anyhow::Result<Vec<u8>>`?

Reply:
I want to have the option to have arbitary IPLD that is visible to the outside world. E.g. if we want to have some hints about the kind of data in the links.

Reply:
and the trait itself makes sure you aren't tied to cbor?

Reply:
I mean if you are just going to use cbor anyway why have `serialize_branch`, `deserialize_branch` and `Link` at all?

Reply:
So Link is mostly for the internal representation. A 32 byte sha2 hash is much smaller than a cid, even if both are `Copy`.

Reply:
Regarding the Ipld thing: I changed this to make it compatible with libipld. We can change it again as long as the on disk format stays the same.